pxtlib/emitter/decompiler.ts

```diff
@@ -304,9 +304,29 @@ ${output}</xml>`;
         }
 
         function emitVariableSetOrChange(name: string, value: ts.Expression | number, changed = false) {
+            const isNumber = typeof value === 'number';
+            if (!isNumber) {
+                const valueNode = value as ts.Expression;
+
+                // Assignments to null are not supported because there is no block value equivalent to null.
+                // However, declarations that initialize to null are fine so long as they occur in the global
+                // scope since blocks that have object types will regenerate the initializations in compilation
+                if (valueNode.kind === SyntaxKind.NullKeyword) {
+                    if (valueNode.parent.kind !== SyntaxKind.VariableDeclaration) {
+                        error(valueNode, "Assigning null to a variable is not supported in blocks");
+                    }
+                    else if (getEnclosingBlockScopeContainer(valueNode).kind !== SyntaxKind.SourceFile) {
+                        error(valueNode, "Only globally scoped initializations to null are supported in blocks");
```

Comment:
It means we can't convert that to blocks (you won't get an error typing that in the editor). That isn't a break from the current behavior, which already doesn't support decompiling anything declared in blocks. Here are some example of what declarations work and what don't:
```typescript
// This code does not decompile, because item is declared in a block
for (let i = 0; i < 4; i++) {
    let item = 0
    item = 1
}

// This code also does not decompile, because item is declared but never referenced
let item = 0
for (let i = 0; i < 4; i++) {
}

// This code does not decompile, because item is not declared in the same place
// as where the compiler would declare it:
input.onButtonPressed(Button.A, () => { })
let item = 0
for (let i = 0; i < 4; i++) {
    item = 1
}

// This code decompiles fine
let item = 0
for (let i = 0; i < 4; i++) {
    item = 1
}
```

Whitespace can mess up decompilation as well. If you place a newline between the declaration and for-loop in my last example, it wouldn't decompile.